caliban_toolbox/aws_functions.py

```diff
@@ -0,0 +1,166 @@
+# prior written permission.
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import sys
+import boto3
+import os
+import threading
+import re
+
+import numpy as np
+from getpass import getpass
+
+from caliban_toolbox.utils.utils import get_img_names, list_npzs_folder
+
+
+# Taken from AWS Documentation
+class ProgressPercentage(object):
+    def __init__(self, filename):
+        self._filename = filename
+        self._size = float(os.path.getsize(filename))
+        self._seen_so_far = 0
+        self._lock = threading.Lock()
+
+    def __call__(self, bytes_amount):
+        with self._lock:
+            self._seen_so_far += bytes_amount
+            percentage = (self._seen_so_far / self._size) * 100
+            sys.stdout.write(
+                "\r%s  %s / %s  (%.2f%%)" % (
+                    self._filename, self._seen_so_far, self._size,
+                    percentage))
+            sys.stdout.flush()
+
+
+def connect_aws():
+    AWS_ACCESS_KEY_ID = getpass('What is your AWS access key id? ')
+    AWS_SECRET_ACCESS_KEY = getpass('What is your AWS secret access key id? ')
+
+    session = boto3.Session(aws_access_key_id=AWS_ACCESS_KEY_ID,
+                            aws_secret_access_key=AWS_SECRET_ACCESS_KEY)
+    print('Connected to AWS')
+    s3 = session.client('s3')
+
+    return s3
+
+
+def aws_upload_files(aws_folder, stage, upload_folder, pixel_only, label_only, rgb_mode):
+    """Uploads files to AWS bucket for use in Figure 8
+
+    Args:
+        aws_folder: folder where uploaded files will be stored
+        stage: specifies stage in pipeline for jobs requiring multiple rounds of annotation
+        upload_folder: path to folder containing files that will be uploaded
+        pixel_only: boolean flag to set pixel_only mode
+        label_only: boolean flag to set label_only mode
+        rgb_mode: boolean flag to set rgb_mode
+    """
+
+    s3 = connect_aws()
+
+    # load the images from specified folder but not the json log file
+    files_to_upload = list_npzs_folder(upload_folder)
+
+    filename_list = []
+
+    # change slashes separating nested folders to underscores for URL generation
+    subfolders = re.split('/', aws_folder)
+    subfolders = '__'.join(subfolders)
+
+    # optional flags to add to URL
+    optional_flags = np.any([pixel_only, label_only, rgb_mode])
+    if optional_flags:
+        optional_url = "?"
+        if pixel_only:
+            optional_url += "&pixel_only=true"
+        if label_only:
+            optional_url += "&label_only=true"
+        if rgb_mode:
+            optional_url += "&rgb=true"
```

Comment:
This clearly works, but I think what you really want to do here is [urlencode your request arguments](https://stackoverflow.com/questions/5607551/how-to-urlencode-a-querystring-in-python). There's a bit of trickyness in getting this to import in 2 and 3, which I think we [resolved in the redis-consumer](https://github.com/vanvalenlab/kiosk-redis-consumer/blob/b3d6c17e8b0f37eb609a39e89f768949fb0900ed/redis_consumer/consumers/base_consumer.py#L680).